### ISSUE TITLE
Rendering mesh to a rectangular image with real camera parameters in PyTorch3D differs from the original image.

### ISSUE THREAD

**Issue by changfali** [facebookresearch/pytorch3d#1956]
Hello, I have scanned data obtained from the MVS method, which includes a mesh and its texture. Additionally, I have the parameters of a specific camera, including the intrinsic matrix K, extrinsic matrix W2C, distortion coefficients k1, k2, k3, p1, p2, and the original image captured by this camera.

I want to render the mesh using this camera and compare the rendered image with the original image. However, the rendered image appears to have a "shift" relative to the original image. Could you help me with this?

[Click this link to view the image](https://pan.baidu.com/s/1ooiECL4OIHmef7JnlHymAg?pwd=x54v ). 

Below is my code(The code might be a bit long, but you can mainly focus on the **_ImageRenderer.PROJECT function_**. I think the main cause of the difference might lie there. 
In the project function, the transformation from normalized coordinates to pixel coordinates is based on the formula at the end of this reference.:[pytorch3d cameras](https://pytorch3d.org/docs/cameras)):
```
import cv2
import torch
import torch.nn as nn
import numpy as np
import torch.nn.functional as F
from pytorch3d.io import load_obj
from pytorch3d.structures import Meshes
from pytorch3d.renderer.mesh import rasterize_meshes


def dict2obj(d):
    if isinstance(d, list):
        d = [dict2obj(x) for x in d]
    if not isinstance(d, dict):
        return d

    class C(object):
        pass

    o = C()
    for k in d:
        o.__dict__[k] = dict2obj(d[k])
    return o


def face_vertices(vertices, faces):
    """
    :param vertices: [batch size, number of vertices, 3]
    :param faces: [batch size, number of faces, 3]
    :return: [batch size, number of faces, 3, 3]
    """
    assert vertices.ndimension() == 3
    assert faces.ndimension() == 3
    assert vertices.shape[0] == faces.shape[0]
    assert vertices.shape[2] == 3
    assert faces.shape[2] == 3

    bs, nv = vertices.shape[:2]
    bs, _ = faces.shape[:2]
    device = vertices.device
    faces = faces + (torch.arange(bs, dtype=torch.int32).to(device) * nv)[:, None, None]
    vertices = vertices.reshape((bs * nv, 3))
    return vertices[faces.long()]


class Pytorch3dRasterizer(nn.Module):
    """
    This class implements methods for rasterizing a batch of heterogenous
    Meshes.
    Notice:
        x,y,z are in image space, normalized
        can only render squared image now
    """
    def __init__(self, image_size=224):
        """
        use fixed raster_settings for rendering faces
        """
        super().__init__()
        raster_settings = {
            "image_size": image_size,
            "blur_radius": 0.0,
            "faces_per_pixel": 1,
            "bin_size": None,
            "max_faces_per_bin": None,
            "perspective_correct": True,
            "cull_backfaces": True,
        }
        raster_settings = dict2obj(raster_settings)
        self.raster_settings = raster_settings
        self.image_size = image_size

    def forward(self, vertices, faces, attributes=None):
        """
        attribute: (N, F, 3, D)
        """
        fixed_vertices = vertices.clone()
        meshes_screen = Meshes(verts=fixed_vertices.float(), faces=faces.long())
        raster_settings = self.raster_settings
        pix_to_face, zbuf, bary_coords, dists = rasterize_meshes(
            meshes_screen,
            image_size=raster_settings.image_size,
            blur_radius=raster_settings.blur_radius,
            faces_per_pixel=raster_settings.faces_per_pixel,
            bin_size=raster_settings.bin_size,
            max_faces_per_bin=raster_settings.max_faces_per_bin,
            perspective_correct=raster_settings.perspective_correct,
        )
        vismask = (pix_to_face > -1).float()

        D = attributes.shape[-1]
        attributes = attributes.clone()
        attributes = attributes.view(attributes.shape[0] * attributes.shape[1], 3, attributes.shape[-1])

        N, H, W, K, _ = bary_coords.shape
        mask = pix_to_face == -1
        pix_to_face = pix_to_face.clone()
        pix_to_face[mask] = 0

        idx = pix_to_face.view(N * H * W * K, 1, 1).expand(N * H * W * K, 3, D)
        pixel_face_vals = attributes.gather(0, idx).view(N, H, W, K, 3, D)
        pixel_vals = (bary_coords[..., None] * pixel_face_vals).sum(dim=-2)
        pixel_vals[mask] = 0  # Replace masked values in output.
        pixel_vals = pixel_vals[:, :, :, 0].permute(0, 3, 1, 2)
        pixel_vals = torch.cat([pixel_vals, vismask[:, :, :, 0][:, None, :, :]], dim=1)
        return pixel_vals, zbuf


class ImageRenderer(nn.Module):
    def __init__(self, verts, faces, aux):
        super(ImageRenderer, self).__init__()
        self.verts = verts
        self.batch_size = verts.shape[0]

        uvcoords = aux.verts_uvs[None, :, :]  # (N, V, 2)   每个顶点在UV贴图中的坐标
        uvfaces = faces.textures_idx[None, :, :]  # (N, F, 3) 每个face上的三个顶点的uvcoords索引
        self.faces = faces.verts_idx[None, :, :]  # （N，F，3）每个face上的三个顶点的序号索引
        vn = aux.normals[None, :, :]

        uvcoords = torch.cat([uvcoords, uvcoords[:, :, 0:1] * 0.0 + 1.0], -1)  # [bz, ntv, 3]
        uvcoords = uvcoords * 2 - 1
        uvcoords[..., 1] = -uvcoords[..., 1]

        self.face_uvcoords = face_vertices(uvcoords, uvfaces)  #
        self.face_normals = face_vertices(vn, self.faces)
        self.attributes = torch.cat(
            [self.face_uvcoords.expand(self.batch_size, -1, -1, -1), self.face_normals.expand(self.batch_size, -1, -1, -1)], -1
        )
        self.tex_img = aux.texture_images[list(aux.texture_images.keys())[0]][None, ::4, ::4, :].permute(0, 3, 1, 2)

    def forward(
        self,
        rasterizer,
        cam,
        image_save_p=None,
    ):

        trans_vertices = self.project(self.verts, cam[0], cam[1], cam[2], rasterizer.image_size[0], rasterizer.image_size[1])

        rendering, zbuf = rasterizer(trans_vertices, self.faces.expand(self.batch_size, -1, -1), self.attributes)

        alpha_images = rendering[:, -1, :, :][:, None, :, :].detach()
        uvcoords_images = rendering[:, :3, :, :]
        grid = (uvcoords_images).permute(0, 2, 3, 1)[:, :, :, :2]

        albedo_images = F.grid_sample(self.tex_img.cuda(), grid, align_corners=False) * alpha_images.repeat(1, 3, 1, 1)

        img = (albedo_images[0].permute(1, 2, 0).cpu().numpy() * 255).astype(np.uint8)[:, :, [2, 1, 0]]
        cv2.imwrite(image_save_p, img)

    def project(self, points, camera_intrinsics, camera_extrinsics, camera_distortions, height, width, eps=1e-7):
        """project the grid points into images (full perspective projection)
        Args:
        - points: (B, N, 3)
        - camera_intrinsics: (B, 3, 4)
        - camera_extrinsics: (B, 3, 3)
        - camera_distortions: (B, 5)
        - height: int
        - width: int

        Return:
        - u_coord: (B, N), value range in [-1, 1]
        - v_coord: (B, N), value range in [-1, 1]
        """

        device = points.device
        batch_size, num_points, _ = points.shape
        points = points.transpose(1, 2).contiguous()

        ones = torch.ones(batch_size, 1, num_points).to(device)
        points_homogeneous = torch.cat((points, ones), axis=-2)  # (batch_size, 4, num_points)

        # Transformation from the world coordinate system to the image coordinate system using the camera extrinsic rotation (R) and translation (T)
        points_image = camera_extrinsics.bmm(points_homogeneous)  # (batch_size, 3, num_points)

        # Transformation from 3D camera coordinate system to the undistorted image plane
        mask = points_image.abs() < eps
        mask[:, :2, :] = False
        points_image[mask] = 1.0  # Avoid division by zero

        points_image_x = points_image[:, 0, :] / points_image[:, 2, :]
        points_image_y = points_image[:, 1, :] / points_image[:, 2, :]
        z__ = points_image[:, 2, :]

        # Transformation from undistorted image plane to distorted image coordinates
        K1, K2, K3, p1, p2 = camera_distortions[:, 0], camera_distortions[:, 1], camera_distortions[:, 2], camera_distortions[:, 3], camera_distortions[:, 4]  # (batch_size)
        r2 = points_image_x**2 + points_image_y**2  # (batch_size, num_points)
        r4 = r2**2
        r6 = torch.pow(r2, 3)
        radial_distortion_factor = 1 + K1[:, None] * r2 + K2[:, None] * r4  + K3[:,None] * r6 # (batch_size, num_points)

        points_image_x = points_image_x * radial_distortion_factor + p1[:,None]*(r2 + 2*points_image_x**2) + 2*p2[:,None]*points_image_x*points_image_y
        points_image_y = points_image_y * radial_distortion_factor + p2[:,None]*(r2 + 2*points_image_y**2) + 2*p1[:,None]*points_image_x*points_image_y
        points_image_z = torch.ones_like(points_image[:, 2, :])
        points_image = torch.cat((points_image_x[:, None, :], points_image_y[:, None, :], points_image_z[:, None, :]), dim=1)

        # Transformation from distorted image coordinates to the final image coordinates with the camera intrinsics
        points_image = camera_intrinsics.bmm(points_image)

        s = min(height, width)

        u_coord = -(points_image[:,0,:]-width/2.0)*2/s
        v_coord = -(points_image[:,1,:]-height/2.)*2/s

        vertices = torch.stack([u_coord, v_coord, z__], dim=-1)
        return vertices


if __name__ == "__main__":
    device = "cuda"
    obj_path = "test_data/scan/5/model.obj"

    verts, faces, aux = load_obj(obj_path, device="cuda")
    # camera_extrinsics
    W2C = torch.tensor([[[ 9.9975e-01,  2.9775e-03, -2.2119e-02, -1.3420e-02],
                         [ 2.9785e-03, -1.0000e+00,  1.1309e-05, -1.1326e-01],
                         [-2.2118e-02, -7.7187e-05, -9.9976e-01, -6.7453e-02],
                         [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  1.0000e+00]]]).float().to(device)
    
    # camera_intrinsics
    K = torch.tensor([[[3.7432e+04, 0.0000e+00, 3.3574e+03],
                       [0.0000e+00, 3.7432e+04, 2.2139e+03],
                       [0.0000e+00, 0.0000e+00, 1.0000e+00]]]).float().to(device)
    
    # camera_distortions
    D = torch.tensor([[ 1.3264e+00,  8.2730e+01, -5.8890e+03, -1.0177e-02, -7.6537e-03]]).float().to(device)
    image_size = [4640,6960]


    render = ImageRenderer(verts[None, :, :], faces, aux)
    rasterizer = Pytorch3dRasterizer(image_size)
    render(rasterizer, [K, W2C, D], "test_render.png")

```
